pkgs/tools/security/pinentry/default.nix

```diff
@@ -1,60 +1,96 @@
 { fetchurl, fetchpatch, stdenv, lib, pkgconfig, autoreconfHook
 , libgpgerror, libassuan
-, libcap ? null, libsecret ? null, ncurses ? null, gtk2 ? null, gcr ? null
-, qt4 ? null, qt5 ? null
-, enableEmacs ? false
+, ncurses, gtk2, qt
+, libcap ? null, libsecret ? null, gcr ? null
+, enabledFlavors ? [ "curses" "tty" "gtk2" "qt" "gnome3" "emacs" ]
 }:
 
-assert qt5 != null -> qt4 == null;
-assert qt4 != null -> qt5 == null;
+with stdenv.lib;
+
+assert isList enabledFlavors && enabledFlavors != [];
 
 let
   mkDerivation =
-    if qt5 != null
-      then qt5.mkDerivation
+    if (builtins.elem "qt" enabledFlavors)
+      then qt.mkDerivation
       else stdenv.mkDerivation;
+
+  mkFlag = pfxTrue: pfxFalse: cond: name:
+    "--${if cond then pfxTrue else pfxFalse}-${name}";
+  mkEnable = mkFlag "enable" "disable";
+  mkWith = mkFlag "with" "without";
+
+  mkEnablePinentry = f:
+    let
+      info = flavorInfo.${f};
+      inputs = info.buildInputs or [];
+      flag = flavorInfo.${f}.flag or null;
+      inputsSatisfied = all (f: ! isNull f) inputs;
+    in
+      assert inputsSatisfied;
```

Comment:
I would rather move this to the start of the expression. Or at least use `assertMsg`.

Reply:
Erm, we can remove that entirely, by removing the `? null` from `gcr`, can we?

We don't set these attributes to `null` anymore, instead use `enabledFlavors`…

Reply:
updated.

Reply:
Actually, `dep ? null` is quite common idiom in nixpkgs. I do not know why we keep cargo culting it even when there are `enableDep` arguments.